kyo-core/shared/src/main/scala/kyo/Resource.scala

```diff
@@ -122,26 +154,11 @@ object Resource:
       *   The result of the effect wrapped in Async and S effects.
       */
     def run[A, S](closeParallelism: Int)(v: A < (Resource & S))(using frame: Frame): A < (Async & S) =
-        Queue.Unbounded.initWith[Unit < (Async & Abort[Throwable])](Access.MultiProducerSingleConsumer) { q =>
-            Promise.initWith[Nothing, Unit] { p =>
-                val finalizer = Finalizer(frame, q)
-                def close: Unit < IO =
-                    q.close.map {
-                        case Absent =>
-                            bug("Resource finalizer queue already closed.")
-                        case Present(tasks) =>
-                            Async.foreach(tasks, closeParallelism) { task =>
-                                Abort.run[Throwable](task)
-                                    .map(_.foldError(_ => (), ex => Log.error("Resource finalizer failed", ex.exception)))
-                            }
-                                .unit
-                                .handle(Async.run[Nothing, Unit, Any])
-                                .map(p.becomeDiscard)
-                    }
-                ContextEffect.handle(Tag[Resource], finalizer, _ => finalizer)(v)
-                    .handle(IO.ensure(close))
-                    .map(result => p.get.andThen(result))
-            }
+        IO.Unsafe {
+            val closeable = Finalizer.Closeable.Unsafe.init(closeParallelism)
+            ContextEffect.handle(Tag[Resource], closeable, _ => closeable)(v)
+                .handle(IO.ensure(closeable.close))
+                .map(result => closeable.await.andThen(result))
```

Comment:
another option is returning a `Fiber` so `use` can be used. Not sure it's worth it